subsys/net/l2_wifi_if_conn/l2_wifi_conn.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(l2_wifi_mgr_conn);
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/wifi_mgmt.h>
+#include <zephyr/net/net_event.h>
+#include <net/wifi_mgmt_ext.h>
+#include <net/l2_wifi_connect.h>
+#include <zephyr/net/conn_mgr_connectivity.h>
+
+#define WIFI_SHELL_MGMT_EVENTS (NET_EVENT_WIFI_CONNECT_RESULT | \
+				NET_EVENT_WIFI_DISCONNECT_RESULT)
+
+static struct net_mgmt_event_callback net_l2_mgmt_cb;
+int connection_status;
+
+
+int net_l2_wifi_connect(struct conn_mgr_conn_binding *const binding);
+int net_l2_wifi_disconnect(struct conn_mgr_conn_binding *const binding);
+
+/* Delayable work used to handle Wi-Fi connection timeouts. */
+static void wifi_conn_work_handler(struct k_work *work);
+K_THREAD_STACK_DEFINE(l2_wifi_wq_stack_area, CONFIG_L2_WIFI_CONN_WQ_STACK_SIZE);
+struct k_work_q wifi_conn_wq;
+static K_WORK_DELAYABLE_DEFINE(wifi_conn_work, wifi_conn_work_handler);
+
+/* Store a local reference to the connection binding */
+struct conn_mgr_conn_binding  *wifi_conn_binding;
```

Comment:
Don't store the binding outside of direct API calls; It's a mutex-protected struct, and accessing outside of callbacks from L2AC bypasses the mutex protection

As it happens, I think you only need to store the iface anyways, not the whole binding, for what you are doing

Reply:
Yes. For most of the parts, iface info will do. We need binding in one case and for that I have made changes to fetch the binding using conn_mgr_if_get_binding call.

Reply:
Unfortunately, `conn_mgr_if_get_binding` is actually a private function that doesn't include the mutex protection for the binding and should not be used except by test suites.

I failed to properly mark it as such, though, I apologize for that confusion. I will make a note of updating its documentation to make this explicit

Fortunately you don't need the binding in that section -- `net_l2_wifi_disconnect` only needs to know the iface, so you can either store that globally or pass **_that_** in as the argument to `net_l2_wifi_disconnect/net_l2_wifi_connect`, so on